File: python_app/services/pantry_service.py
import abc
import dataclasses
import logging
import uuid
from typing import List

logger = logging.getLogger(__name__)

@dataclasses.dataclass
class LineItem:
    name: str
    amount: float
    unit: str

@dataclasses.dataclass
class PantryEntry:
    id: uuid.UUID = dataclasses.field(default_factory=uuid.uuid4)
    name: str
    amount: float
    unit: str

@dataclasses.dataclass
class StorageRequest:
    items: List[LineItem]

@dataclasses.dataclass
class UseFoodRequest:
    items: List[LineItem]

class DurablePantry(abc.ABC):
    @abc.abstractmethod
    def find_all(self) -> List[PantryEntry]:
        pass

    @abc.abstractmethod
    def find_all_where_names_exist(self, names: List[str]) -> List[PantryEntry]:
        pass

    @abc.abstractmethod
    def save_all(self, entries: List[PantryEntry]) -> List[PantryEntry]:
        pass

class PantryService:
    def __init__(self, durable_pantry: DurablePantry):
        self.durable_pantry = durable_pantry

    def get_food(self) -> List[PantryEntry]:
        logger.info("Finding all pantry entries")
        return self.durable_pantry.find_all()

    def save_food(self, request: StorageRequest) -> List[PantryEntry]:
        logger.info(f"Saving food items: {request.items}")
        
        names_to_find = [item.name for item in request.items]
        existing_entries = self.durable_pantry.find_all_where_names_exist(names_to_find)
        
        existing_map = {entry.name: entry for entry in existing_entries}
        
        updated_entries = []
        new_entries = []

        for item in request.items:
            if item.name in existing_map:
                entry = existing_map[item.name]
                if entry.unit != item.unit:
                    logger.error(f"Unit mismatch for {item.name}: existing {entry.unit}, new {item.unit}")
                    raise ValueError(f"Unit mismatch for {item.name}")
                entry.amount = item.amount # Changed from += to =
                updated_entries.append(entry)
            else:
                new_entries.append(PantryEntry(name=item.name, amount=item.amount, unit=item.unit))
        
        saved_entries = self.durable_pantry.save_all(updated_entries + new_entries)
        logger.info(f"Saved entries: {saved_entries}")
        return saved_entries

    def use_food(self, request: UseFoodRequest) -> List[PantryEntry]:
        logger.info(f"Using food items: {request.items}")

        names_to_find = [item.name for item in request.items]
        existing_entries = self.durable_pantry.find_all_where_names_exist(names_to_find)
        
        existing_map = {entry.name: entry for entry in existing_entries}
        
        entries_to_save = []

        for item in request.items:
            if item.name not in existing_map:
                logger.error(f"Item not found in pantry: {item.name}")
                # Removed: raise ValueError(f"Item not found: {item.name}")
                continue # Skip to the next item

            entry = existing_map[item.name]
            if entry.unit != item.unit:
                logger.error(f"Unit mismatch for {item.name}: existing {entry.unit}, requested {item.unit}")
                raise ValueError(f"Unit mismatch for {item.name}")
            
            if entry.amount < item.amount:
                logger.error(f"Not enough {item.name} in pantry: requested {item.amount}, available {entry.amount}")
                raise ValueError(f"Not enough {item.name}")

            entry.amount -= item.amount
            entries_to_save.append(entry)
            
        saved_entries = self.durable_pantry.save_all(entries_to_save)
        logger.info(f"Updated entries after use: {saved_entries}")
        return saved_entries

File: src/main/kotlin/stift/io/home/domain/PantryService.kt
package stift.io.home.domain

import dev.langchain4j.agent.tool.Tool
import java.util.*
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

// TODO: We want the LLM to be able to use methods on this class. How can we provide them? (Hint: @Tool())
// TODO: Assuming we had User Entities and Spring Security - how could we retain the UserContext on a path through an LLM?
// TODO: How can we do that safely?
interface DurablePantry {
    fun findAll(): List<PantryService.PantryEntry>
    fun findAllWhereNamesExist(
        names: List<String>
    ): List<PantryService.PantryEntry>

    fun saveAll(entries: List<PantryService.PantryEntry>): List<PantryService.PantryEntry>
}

@Service
class PantryService(
    private val durablePantry: DurablePantry,
) {
    data class LineItem(val name: String, val amount: Double, val unit: String)

    data class StorageRequest(val items: List<LineItem>)

    data class UseFoodRequest(val items: List<LineItem>)

    data class PantryEntry(
        val id: UUID,
        val name: String,
        val amount: Double,
        val unit: String
    )

    fun saveFood(request: StorageRequest): List<PantryEntry> {
        logger.info("Incoming save food Request: $request")
        val saveRequestList = request.items
        val existingEntries =
            durablePantry.findAllWhereNamesExist(request.items.map { it.name })

        val updateList =
            saveRequestList.map { entryToSave ->
                existingEntries
                    .find { it.name == entryToSave.name }
                    ?.copy(amount = entryToSave.amount, unit = entryToSave.unit)
                    ?: PantryEntry(
                        id = UUID.randomUUID(),
                        name = entryToSave.name,
                        amount = entryToSave.amount,
                        unit = entryToSave.unit
                    )
            }

        val savedEntries = durablePantry.saveAll(updateList)

        return savedEntries
    }

    fun useFood(request: UseFoodRequest): List<PantryEntry> {
        logger.info("Incoming use food Request: $request")
        val existingEntries =
            durablePantry.findAllWhereNamesExist(request.items.map { it.name })

        val updatedEntries =
            request.items.mapNotNull { itemToUse ->
                existingEntries
                    .find { it.name == itemToUse.name }
                    ?.let { existingItem ->
                        val newAmount = maxOf(0.0, existingItem.amount - itemToUse.amount)
                        existingItem.copy(amount = newAmount)
                    }
                    ?.also {
                        if (it.amount == 0.0) {
                            logger.info("Item ${it.name} is now depleted")
                        }
                    }
                    ?: run {
                        logger.error("Item not in storage: ${itemToUse.name}")
                        null
                    }
            }

        val savedEntries =
            if (updatedEntries.isNotEmpty()) {
                durablePantry.saveAll(updatedEntries)
            } else {
                emptyList()
            }

        return savedEntries
    }

    fun getFood(): List<PantryEntry> {
        return durablePantry.findAll()
    }

    companion object {
        private val logger = LoggerFactory.getLogger(this::class.java)
    }
}
